src/nf/nf_optimizers.f90

```diff
@@ -123,4 +150,45 @@ pure subroutine minimize_rmsprop(self, param, gradient)
 
   end subroutine minimize_rmsprop
 
+
+  impure elemental subroutine init_adam(self, num_params)
+    class(adam), intent(inout) :: self
+    integer, intent(in) :: num_params
+    if (.not. allocated(self % m)) then
+      allocate(self % m(num_params), self % v(num_params))
+      self % m = 0
+      self % v = 0
+    end if
+  end subroutine init_adam
+
+
+  pure subroutine minimize_adam(self, param, gradient)
+    !! Concrete implementation of an Adam optimizer update rule.
+    class(adam), intent(inout) :: self
+    real, intent(inout) :: param(:)
+    real, intent(in) :: gradient(:)
+
+    self % t = self % t + 1
+
+    ! If weight_decay > 0, use L2 regularization;
+    ! otherwise, default to regular Adam.
+    associate(g => gradient + 2 * self % weight_decay * param)
+      self % m = self % beta1 * self % m + (1 - self % beta1) * g
+      self % v = self % beta2 * self % v + (1 - self % beta2) * g**2
+    end associate
+
+    ! Compute bias-corrected first and second moment estimates.
+    associate( &
+      m_hat => self % m / (1 - self % beta1**self % t), &
+      v_hat => self % v / (1 - self % beta2**self % t) &
+    )
+
+      ! Update parameters.
+      param = param - self % learning_rate &
+        * m_hat / (sqrt(v_hat) + self % epsilon)
```

Comment:
Following up on my comment above, should the decoupled weight regularization not be:

```suggestion
      param = param - self % learning_rate &
        * m_hat / (sqrt(v_hat) + self % epsilon) &
        - self % weight_decay * param
```

Reply:
Yeah, have made that change, just pushed